feat: allows configure to consider parent folder name in queries

cristianoliveira opened the pull request:
Hello there :wave: 

I took the liberty to implement a feature that was missing when filtering bookmarks. Usually I organize my bookmarks in folders like `projects`, `tools`, etc  and when querying the bookmarks I had to query by their exact name. I wanted to filter them but "pre filtered" by the parent name first.

Like this:
```
Bookmarks/
  Github home 
  python/
      Learn Python - Getting started tutorial
      advanced/
        Python Tricks - Pro tutorial
 ```
 
 I should be able to query by `tutorial` and receive both suggestions, as well as, query by `advanced tutorial` and receive only one (see the unit tests)

I'm not sure if this is align with the direction of the plugin, feel free to close the PR if it isn't

## Extra
 
 I also added some tests and a Makefile to run then with `make test`

cristianoliveira committed:
feat: consider parents name in queries

cristianoliveira committed:
fix: comply to linter

cristianoliveira commented:
![crop](https://github.com/user-attachments/assets/843b942b-8689-469c-b0ce-57312c99f04e)

pascalbe-dev commented:
Hey @cristianoliveira , 

thanks for the input and sorry for the late reply. I've been quite busy recently. 

This looks interesting. I do not have this use case for myself, but it does make sense and I could imagine more people who organize it this way. If it's non breaking (people without such a structure can still use it like now), I do not see a problem in applying this change. 

I'll take a deeper look within the next days (probably monday). Thanks for the unit tests also. Makes total sense and I have not gotten into that here, so it's great to a first part. Awesome. 

In the meantime (if you have not already), you can fork and then just use your fork as repo. I'll keep you updated here once there is news to share (and you'll see it anyway if I merge). 

Thanks again and have a nice week 💪🏼

cristianoliveira commented:
Hey @pascalbe-dev keeping backwards compatibility makes sense! I was thinking to add a configuration to apply this behaviour

cristianoliveira committed:
feat: adds 'filter_by_folders' setting

Summary:
This commit adds a feature to filter bookmarks by folder structure during search queries

Detailed Changes:
 - BrowserBookmarks.py:
   - Modified `BookmarkQuerier` instantiation to include a `filter_by_folders` parameter from preferences.
 - manifest.json:
   - Added a new setting `filter_by_folders` with select options for enabling/disabling folder filtering.
 - querier/BookmarkQuerier.py:
   - Updated `BookmarkQuerier` class to accept `filter_by_folders` parameter and modify search behavior accordingly.
 - tests/test_bookmark_querier.py:
   - Split tests into separate classes for when `filter_by_folders` is enabled and disabled. Added new test cases to ensure correct functionality when the filter is applied.

(cherry picked from commit 11c62fcf114dd5fc537843e790b8f345ac50ca60)

cristianoliveira commented:
Added a new setting `filter_by_folder` to enable/disable the feature, the default is disabled :D

cristianoliveira committed:
tests: remove duplication

cristianoliveira committed:
chore: remove dead code

cristianoliveira committed:
chore: apply lint rules

cristianoliveira committed:
docs: remove yes/no description in favor of selector

cristianoliveira committed:
feat: matches parent tree

cristianoliveira committed:
tests: split tests by checks

cristianoliveira committed:
tests: use one fixture to show both behaviours

pascalbe-dev commented:
@cristianoliveira 

Finally merged. Sorry for the long delay. Thanks for your contribution :)
